test/e2e/azure_vmextensions.go

```diff
@@ -0,0 +1,112 @@
+
+	input = inputGetter()
+	Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+	Expect(input.Namespace).ToNot(BeNil(), "Invalid argument. input.Namespace can't be nil when calling %s spec", specName)
+	Expect(input.ClusterName).ToNot(BeEmpty(), "Invalid argument. input.ClusterName can't be empty when calling %s spec", specName)
+
+	By("creating a Kubernetes client to the workload cluster")
+	workloadClusterProxy := input.BootstrapClusterProxy.GetWorkloadCluster(ctx, input.Namespace.Name, input.ClusterName)
+	Expect(workloadClusterProxy).NotTo(BeNil())
+	mgmtClient := bootstrapClusterProxy.GetClient()
+	Expect(mgmtClient).NotTo(BeNil())
+
+	By("Retrieving all machines from the machine template spec")
+	machineList := &infrav1.AzureMachineList{}
+	// list all of the requested objects within the cluster namespace with the cluster name label
+	Logf("Listing machines in namespace %s with label %s=%s", input.Namespace.Name, clusterv1.ClusterLabelName, workloadClusterProxy.GetName())
+	err := mgmtClient.List(ctx, machineList, client.InNamespace(input.Namespace.Name), client.MatchingLabels{clusterv1.ClusterLabelName: workloadClusterProxy.GetName()})
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Creating a mapping of machine IDs to array of expected VM extensions")
+	expectedExtensionMap := make(map[string][]string)
+	for _, machine := range machineList.Items {
+		for _, extension := range machine.Spec.VMExtensions {
+			expectedExtensionMap[*machine.Spec.ProviderID] = append(expectedExtensionMap[*machine.Spec.ProviderID], extension.Name)
+		}
+	}
+
+	By("Creating a VM and VM extension client")
+	// get subscription id
+	settings, err := auth.GetSettingsFromEnvironment()
+	Expect(err).NotTo(HaveOccurred())
+	subscriptionID := settings.GetSubscriptionID()
+	auth, err := settings.GetAuthorizer()
+	Expect(err).NotTo(HaveOccurred())
+
+	// create a VM client
+	vmClient := compute.NewVirtualMachinesClient(subscriptionID)
+	vmClient.Authorizer = auth
+
+	// create a VM extension client
+	extensionsClient := compute.NewVirtualMachineExtensionsClient(subscriptionID)
+	extensionsClient.Authorizer = auth
+
+	vmListResults, err := vmClient.List(ctx, input.ClusterName, "")
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Verifying specified VM extensions are created on Azure")
+	for _, machine := range vmListResults.Values() {
+		vmExtensionListResult, err := extensionsClient.List(ctx, input.ClusterName, *machine.Name, "")
+		Expect(err).NotTo(HaveOccurred())
+		vmExtensionList := *vmExtensionListResult.Value
+		var vmExtensionNames []string
+		for _, vmExtension := range vmExtensionList {
+			vmExtensionNames = append(vmExtensionNames, *vmExtension.Name)
+		}
+		osName := string(machine.VirtualMachineProperties.StorageProfile.OsDisk.OsType)
+		Expect(vmExtensionNames).To(ContainElements("CAPZ." + osName + ".Bootstrapping"))
```

Comment:
So one thing that occurs to me: because the bootstrapping extension is present on (non-AKS) clusters we build, we could run this validation on all (non-AKS) cluster scenarios. It looks like you've already wired this up to detect the extensions defined on the capi resources, so we shouldn't have to curate which cluster scenarios we invoke `AzureVMExtensionsSpec` against.

Reply:
That's true, should I added it to all other E2E test scenarios besides the AKS one? Theoretically this should work with other test scenarios as well, as it checks for extensions that are defined in the spec.